### PR TITLE
Add translation support

### PR DESCRIPTION
**Pull request by simon04**
See #29.

Required for https://josm.openstreetmap.de/ticket/18140

### CODE REVIEW

**Comment by simonpoole** [comment]
Thanks for working on this. 

I will need some convincing that this level of indirection, that is supporting external translations/translations mechanisms, is actually required or even a good thing.

Assume we merge this, then the next thing I would do is to add a default implementation so that the library can be used self contained with its own translations, which is what IMHO most devs would expect and I'm not quite sure what added value there is, even for the JOSM project, in having separate translations.  

I do note that we would need to provide a way to specify the target Locale, as there are a number of cases in which you might want to deviate from the system default.

**Comment by simon04** [comment]
Fair enough. For Java, the standard way would be using https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ResourceBundle.html, JOSM follows the gettext approach, and for Android the situation is different again: https://developer.android.com/guide/topics/resources/localization.html – Android seems to support ResourceBundle: https://developer.android.com/reference/java/util/ResourceBundle

What do you propose? Should we follow the ResourceBundle approach? Would you push the strings to Transifex for translation?

**Comment by simonpoole** [comment]
> What do you propose? Should we follow the ResourceBundle approach? 

I would like to, if at all possible, try to keep the library 3rd party dependency free, so this would be the obvious solution. I would just want to test if it actually works on Android (always sceptical about non-mainstream things), will give it a look tomorrow.

> Would you push the strings to Transifex for translation?

Yes, we have lots of translators for numerous OSM related projects there https://www.transifex.com/openstreetmap/public/  so it makes most sense. The format is supported https://docs.transifex.com/formats/java-properties so that looks good.

**Comment by simon04** [comment]
I've implemented translation support using Java's ResourceBundle.

**Comment by simonpoole** [comment]
OK, the good news using ResourceBundle actually works on Android.

However we have an issue in that the messages created by the generated parser are not/can't be translated. As there is no non-hacky way to stop JavaCC from overwriting ParseException.java using a modified ParseException class would be ugly. My preference instead would be to wrap the rules method in the javacc input itself. 

Roughly as in

```
public class OpeningHoursParser
{
  static int TWELVEHOURS = 12*60;
  
  boolean strict = false;
  enum AMPM { AM, PM };
  boolean monthOnly = false;

  List<Rule > getRules(boolean strict) throws OpeningHoursParseException {
    try {
      return rules(strict);
    } catch (OpeningHoursParseException ohpex) {
      throw ohpex;
    } catch (ParseException pex) {
      throw new OpeningHoursParseException(pex);
    }
  }
}
```
The downsides of this is that we introduce a new method to get the parse results (but in a backwards compatible fashion) and that we will need to duplicate some code from ParseException in OpeningHoursParseException (essentially the _initialise_ method).

Better/alternative suggestions welcome :-)

File: src/main/java/ch/poole/openinghoursparser/OpeningHoursParseException.java [hunk 1]
```diff
@@ -43,7 +46,7 @@ public int getColumn() {
     @Override
     public String getMessage() {
```
Comment (simonpoole) [inline src/main/java/ch/poole/openinghoursparser/OpeningHoursParseException.java]: Shouldn't we be using _getLocalizedMessage_ to return translated messages and just leave _getMessage_ as is? I admit to never using the former though too, so maybe this is the wrong place for Java purity :-).r

**Comment by simon04** [comment]
> As there is no non-hacky way to stop JavaCC from overwriting ParseException.java using a modified ParseException class would be ugly.

I'm confused. I only find `ch.poole.openinghoursparser.ParseException` which is generated from JavaCC. I've also inspected the generated JAR file and couldn't find any other subclass besides our `ch.poole.openinghoursparser.OpeningHoursParseException`. In the SDK, there's also `java.text.ParseException`, but this is nowhere used. Am I missing something? Or are you referring the the message generated in `ch.poole.openinghoursparser.ParseException#initialise`?

**Comment by simonpoole** [comment]
> Or are you referring the the message generated in `ch.poole.openinghoursparser.ParseException#initialise`?

Yes. Right now we would have the inconsistent situation that the exceptions thrown "manually" mainly in strict made would be translated, but those thrown "automatically" are not.

**Comment by simon04** [comment]
Here we go…

**Comment by simonpoole** [comment]
Looks good (had my brain in neutral wrt _rules()_). Final point: did you see my comment on _getMessage_ vs _getLocalizedMessage_ I have no strong feelings on the matter.

**Comment by simon04** [comment]
Only implementing `getLocalizedMessage` and leaving the default `getMessage` will provide incomplete information (e.g. no line numbers) to users of `getMessage`. Implementing both seems to be wasted effort. So, I'd stick with the current implementation.

**Comment by simonpoole** [comment]
Fine with me. Merging in a sec.

**Comment by simon04** [comment]
Hijacking this issue to say thank you! Some hours ago, I successfully integrated OpeningHoursParser in JOSM: https://josm.openstreetmap.de/changeset/15978/josm

Now, we've reduced the number of JavaScript ScriptEngine usages in JOSM from 3 to 2, and spare 10 MB of parsed JavaScript code on the heap.

Thanks again!

**Comment by simonpoole** [comment]
NP, glad to be of help.

**Pull request merged**